File: src/main/java/leetcode/offer/Offer26.java
package leetcode.offer;

/**
 * @author jingxinwu
 * @date 2021-11-21 6:22 下午
 */
public class Offer26 {
    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if (A == null && B == null) {
            return true;
        }
        if (A == null) {
            return false;
        }
        if (B == null) {
            return true;
        }

        if (A.val == B.val) {
            return true;
        }
        return isSubStructure(A.left, B.left) || isSubStructure(A.right, B.right);
    }
}

File: go/src/leetcode/offer/Offer26.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 18:52
 */

func isSubStructure(A *TreeNode, B *TreeNode) bool {
    if A == nil || B == nil {
        return false
    }
	var dfs func(A *TreeNode, B *TreeNode) bool
	dfs = func(A *TreeNode, B *TreeNode) bool {
	    if (A == null && B == null) {
            return true;
        }
		if A == nil {
			return false
		}
        if B == nil {
            return true
        }
        if A.Val != B.Val {
            return false
        }
		return dfs(A.Left, B.Left) && dfs(A.Right, B.Right)
	}
	return dfs(A, B) || isSubStructure(A.Left, B) || isSubStructure(A.Right, B)
}